### DISCUSSION TITLE
Is strongSwan FIPS compliant on Rocky Linux?

### DISCUSSION THREAD

**Discussion by JimChen111** [strongswan/strongswan#2253] [Q&A] [answered]
Hi,

We are planning to use strongSwan to set up the IPSec tunnels in our project. The OS we are using is Rocky Linux. And we are looking into the FIPS compliance of strongSwan on Rocky but we didn't find much documentation for it.

StrongSwan is FIPS certified on Ubuntu from this [NIST page](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/2978). But for Rocky Linux, based on [CIQ Rocky FIPS](https://github.com/ciq-rocky-fips), strongSwan is not included as a FIPS certified package. But in the meanwhile, the kernel and OpenSSL that we are using are FIPS certified.

So now we wonder if strongSwan is still FIPS compliant on Rocky. We have a couple of questions on this:

1. We noticed that strongSwan has an [option](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html#_charon_plugins_openssl) to enable FIPS mode. If we enable it in our environment (we have FIPS compliant Rocky kernel and OpenSSL), then does it mean the strongSwan we are using on Rocky will become FIPS compliant? (Or it only applies to Ubuntu?)
2. I also found this [post](https://github.com/strongswan/strongswan/discussions/1716?new_signup=true?new_signup=true) about strongSwan FIPS on RHEL. I think the answer mentioned it was possible to make the strongSwan on RHEL FIPS compliant by disabling the default crypto plugins. Could you elaborate more on that? It'll be better if we can have a list of steps to follow to ensure the strongSwan on RHEL/Rocky to be FIPS compliant.

Thanks in advance!

**Comment by tobiasbrunner** [comment, marked as answer]
> 1. We noticed that strongSwan has an [option](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html#_charon_plugins_openssl) to enable FIPS mode. If we enable it in our environment (we have FIPS compliant Rocky kernel and OpenSSL), then does it mean the strongSwan we are using on Rocky will become FIPS compliant? (Or it only applies to Ubuntu?)

That just enforces FIPS-mode for the _openssl_ plugin or rather OpenSSL's libcrypto library. Which is a prerequisite for FIPS-compliance I'd assume. That's platform-independent but doesn't affect strongSwan in any other way.

> 2\. I also found this [post](https://github.com/strongswan/strongswan/discussions/1716?new_signup=true?new_signup=true) about strongSwan FIPS on RHEL. I think the answer mentioned it was possible to make the strongSwan on RHEL FIPS compliant by disabling the default crypto plugins. Could you elaborate more on that? It'll be better if we can have a list of steps to follow to ensure the strongSwan on RHEL/Rocky to be FIPS compliant.

As I wrote there, disabling the default crypto plugins in strongSwan 5.9.6+ and enabling one that links a FIPS-certified crypto library (e.g. _openssl_) prevents strongSwan from doing any crypto operations itself, which should simplify FIPS certification.  You can disable all the default plugins and options via `--disable-defaults`. However, you then have to enable even basic stuff like `--enable-ikev2`. You could also disable the default crypto plugins manually via their `--disable-*` options (see the [autoconf options](https://docs.strongswan.org/docs/5.9/install/autoconf.html) or, even better, check `./configure --help` for your specific version). The script will print a summary of the enabled plugins at the end. Make sure you don't have any disabled plugins installed from a previous build.

**Reply by JimChen111** [reply]
Thanks Tobias! I have some follow up questions for this:

- Does this only apply to the strongSwan after (including) version 5.9.6?
- I suppose we should then only enable openssl (`--enable-openssl`) and disable all other crypto plugins, is that correct? Then how do we make sure that we ARE disabling every other crypto plugin needed? Is there a fast way to get the list or we do need to go through the list one by one and learn if it's crypto related?
- We haven't built/installed our strongSwan from source code before. Is there a way to learn what we have enabled in our current installed version without `./configure`?

Thanks again!

**Reply by tobiasbrunner** [reply]
> * Does this only apply to the strongSwan after (including) version 5.9.6?

Yes, previous versions did the IKEv2 key derivation internally (`prf+()`, which is basically the code that's now in the _kdf_ plugin as a fallback if it's not handled by a third-party library via HKDF). You can have a look at the [Ubuntu FIPS certificates](https://ubuntu.com/security/fips) for older strongSwan version, which explicitly cover the IKEv2 KDF.



> * I suppose we should then only enable openssl (`--enable-openssl`) and disable all other crypto plugins, is that correct?

Yes, pretty much. Unless you start from scratch and use `--disable-defaults` and then only enable what you really need.

> Is there a fast way to get the list or we do need to go through the list one by one and learn if it's crypto related?

Yeah, I guess. At least the list of the enabled plugins is available (`./configure` prints it at the end, you can also find it in `config.log`), so you know where to start if you go the `--disable-*` route (however, that could be tricky when updating to newer versions as the defaults might change, they definitely will with 6.0). Which plugins you need, really depends on your use case.

> * We haven't built/installed our strongSwan from source code before. Is there a way to learn what we have enabled in our current installed version without `./configure`?

The loaded plugins are listed when the daemon is started (tools have a `--help` option that often prints the loaded plugins), other than that, you'd have to check the package sources (e.g. `debian/rules` or a `.spec` file).

**Reply by JimChen111** [reply]
Thanks for the quick reply again!

Then one more question I have is, do all the plugins enabled by default have their 'replacements' supported in OpenSSL? If we accidentally disabled something basic that we weren't supposed to, will any errors be reported at build time or runtime?

**Reply by tobiasbrunner** [reply]
> Then one more question I have is, do all the plugins enabled by default have their 'replacements' supported in OpenSSL?

Pretty much (see [the docs](https://docs.strongswan.org/docs/5.9/config/IKEv2CipherSuites.html)), but there are two AES-based PRFs that are only supported via _kdf_ and _cmac_ and _xcbc_ plugins, but they are probably not used very often.

> If we accidentally disabled something basic that we weren't supposed to, will any errors be reported at build time or runtime?

You'll definitely notice at runtime if something is not supported.

**Reply by JimChen111** [reply]
Wonderful. I also found this [example](https://docs.strongswan.org/docs/5.9/install/autoconf.html#_example) from the option doc. It says the 

`
example builds a strongSwan IKEv2 [charon-systemd](https://docs.strongswan.org/docs/5.9/install/autoconf.html#daemons/charon-systemd) daemon supporting the authentication methods pubkey, psk, eap-md5 and eap-tls. All crypto functions are based on the openssl plugin.
`

So here, can we also delegate the authentication and certification to OpenSSL? Or we can only do for cryptos? E.g. can we even disable things like `psk`, `pem`, `pgp`, `pubkey`, `agent`, etc...?

**Reply by tobiasbrunner** [reply]
> So here, can we also delegate the authentication and certification to OpenSSL? Or we can only do for cryptos?

Authentication is crypto (i.e. signatures), or what do you mean?

> E.g. can we even disable things like `psk`, `pem`, `pgp`, `pubkey`, `agent`, etc...?

There is no _psk_ plugin/feature, that's always supported, the _pem_ plugin could be necessary but it depends (the _openssl_ plugin can load some data directly in PEM encoding but not all, if you provide everything in DER encoding it's not necessary), you most likely don't need the _pgp_ plugin (these type of keys/certificates are rarely used), the _pubkey_ plugin is only needed for plain public keys (wraps them so they can be used like certificates) and the _agent_ plugin you could enable if you wanted to delegate the signature creation to an SSH or GPG agent. Some of the other stuff enabled in that example you might not need (e.g. EAP or TPM), you should check the lists of [options](https://docs.strongswan.org/docs/5.9/install/autoconf.html) and [plugins](https://docs.strongswan.org/docs/5.9/plugins/plugins.html) and decide for yourself.
